### PR TITLE
Onnx backend: perch2 and birdnet, move tensorflow dependencies as optional

### PR DESCRIPTION
**Pull request by nkundiushuti**
Replace birdnetlib/TFLite with ONNX runtime; add Perch v2

  Summary

  - New model perch2: Google Perch v2 via ONNX (justinchuby/Perch-onnx), 14 795 species, 1 536-d embeddings, 32 kHz. No TensorFlow dependency.
  - BirdNet rewritten: avex/models/birdnet.py now runs against justinchuby/BirdNET-onnx via onnxruntime. Removes the
  birdnetlib and TFLite stack entirely. The ONNX graph is patched at load time to expose the 1024-d penultimate embedding tensor as a second output; falls back to the unoptimised model.onnx if shape information was stripped by onnxslim.
  - TensorFlow made optional: tensorflow and tensorflow-hub moved to avex[tensorflow]. birdnetlib removed from core
  dependencies. onnx and onnxruntime added as core dependencies.
  - librosa made optional: with birdnetlib gone, nothing in the package imports librosa anymore (only examples/*.py), so it moves to a new avex[examples] extra. scikit-learn is now declared as an explicit core dependency — it is imported directly by avex.evaluation/avex.metrics/avex.data.transforms but was previously only pulled in transitively via librosa.
  - Python 3.10 dropped (requires-python >=3.11,<3.14): onnxruntime >=1.24 no longer ships cp310 wheels, and 3.10 is near end-of-life.
  - Deprecation warnings: perch (Perch v1) and surfperch emit DeprecationWarning on instantiation pointing users to perch2 and avex[tensorflow].
  - Docs updated: supported_models.md, INSTALL.md, and README.md reflect the new ONNX models, the deprecated TF models, and the optional extra install.

  Migration validation

  Validated the ONNX BirdNet against the previous birdnetlib/TFLite model by installing `avex==1.2.0` in an isolated env, capturing reference outputs on repo audio fixtures, and comparing against this branch's ONNX model (CPU).

  BirdNET always runs at 48 kHz internally; inputs are resampled to 48 kHz before inference. The "source SR" column below is the sample rate of the input fixture, i.e. whether it exercises the resample path — not a model mode.

  **Numerical parity** — mean-aggregated 1024-d embeddings, 5 s clips:

  | Fixture | Source SR (resampled to 48 kHz) | Embedding cosine vs v1.2.0 | Top-1 logit class match |
  |---------|:-------------------------------:|:--------------------------:|:-----------------------:|
  | insectset (Diceroprocta) | 48 kHz (no resample) | 1.000000 (L2 0.015) | ✅ |
  | xeno_canto turtur (XC564654) | 16 kHz → 48 kHz | 0.997729 | ✅ |
  | animalspeak iNat 246886 | 16 kHz → 48 kHz | 0.994910 | ✅ |

  Embedding cosine ≥ 0.9949 and top-1 detection class agrees on every fixture. The already-48 kHz fixture is essentially identical (cos 1.0); the small gaps on the 16 kHz sources come from the pre-model resampling path differing (old librosa vs new torchaudio) before the audio reaches the model — the model step itself is identical.

  The low logit Pearson r (≈0.1) is expected and not a regression: the original forward() applies sigmoid and returns a sparse birdnetlib detection-confidence vector (most entries zero), while the ONNX forward() returns dense raw logits.

  **Latency** — CPU, 5 s clip, warmup 3 + 20 runs (mean). Source SR = input fixture rate (both resampled to 48 kHz before inference):

  | Metric | Old (birdnetlib/TFLite) | New (ONNX) | Speedup |
  |--------|------------------------:|-----------:|:-------:|
  | Model load | 4586 ms | 963 ms | 4.8× |
  | extract_embeddings — 16 kHz source | 82.97 ms | 24.83 ms | 3.3× |
  | extract_embeddings — 48 kHz source | 88.56 ms | 21.40 ms | 4.1× |
  | forward (logits) — 16 kHz source | 38.28 ms | 32.26 ms | 1.2× |
  | forward (logits) — 48 kHz source | 41.90 ms | 16.84 ms | 2.5× |

  ONNX is faster on every axis. Old embedding extraction was disproportionately slow because the TFLite path enables experimental_preserve_all_tensors to reach the penultimate layer; the ONNX graph exposes the embedding as a second output, so extraction ≈ forward cost.

### CODE REVIEW

File: avex/models/birdmae.py [hunk 1]
```diff
@@ -58,7 +58,7 @@ def __init__(
             from transformers import AutoFeatureExtractor, AutoModel
         except ImportError as e:
             raise ImportError(
-                "transformers library is required for BirdMAE. Install with: pip install transformers"
+                "transformers library is required for BirdMAE. Install with: uv pip install transformers"
```
Comment (GaganNarula) [inline avex/models/birdmae.py]: we can just say `transformers` library is required and the user can use their preferred package manager

File: avex/models/birdnet.py [hunk 2]
```diff
@@ -1,574 +1,408 @@
...
+    from huggingface_hub import hf_hub_download
 
-    SAMPLE_RATE = 48_000
-    CHUNK_SEC = 3.0  # fixed window length inside the model
+    try:
+        import onnx as onnx_lib
+        import onnx.shape_inference
+        import onnxruntime as ort
+    except ImportError as e:
+        raise ImportError("onnxruntime and onnx are required.\nuv pip install onnxruntime onnx") from e
```
Comment (GaganNarula) [inline avex/models/birdnet.py:113]: onnx is now in the main dependencies isnt it ? so why wouldnt it be installed ? i'd move the import to module top

File: avex/models/birdnet.py [hunk 3]
```diff
@@ -1,574 +1,408 @@
...
+    def _try_load(filename: str) -> tuple[Any, str, str, str | None]:
+        model_path = hf_hub_download(repo_id=repo_id, filename=filename)
+        logger.info("Loading BirdNet ONNX: %s", model_path)
+
+        model_proto = onnx_lib.load(model_path)
+        model_proto = onnx.shape_inference.infer_shapes(model_proto)
+
+        inp_name = model_proto.graph.input[0].name
+        clf_out_name = model_proto.graph.output[0].name
+        emb_name = _find_embedding_tensor_name(model_proto)
```
Comment (GaganNarula) [inline avex/models/birdnet.py:126]: i dont quite understand this function.. what is it searching for ? why are we searching for a 1024 dim embedding tensor _only_ ?

File: avex/models/birdnet.py [hunk 4]
```diff
@@ -1,574 +1,408 @@
...
+
+    while queue:
+        name = queue.pop(0)
+        if name in visited:
+            continue
+        visited.add(name)
+
+        shape = name_to_shape.get(name, [])
+        # Match [batch, 1024] or [1024] — 2-D or 1-D with correct last dim
+        if _EMBEDDING_DIM in shape and len(shape) <= 2:
```
Comment (GaganNarula) [inline avex/models/birdnet.py:77]: shouldn't we check `shape[-1] == _EMBEDDING_DIM` ?

File: avex/models/birdnet.py [hunk 5]
```diff
@@ -1,574 +1,408 @@
...
-        elif aggregation == "max":
-            # Max pooling over time dimension for each clip
-            result = torch.stack([emb.max(0)[0] for emb in batch_out])  # (B, 1024)
-        elif aggregation == "cls_token":
-            # For BirdNET, treat as mean since no CLS token
-            result = torch.stack([emb.mean(0) for emb in batch_out])  # (B, 1024)
-        else:
-            raise ValueError(f"Unsupported aggregation method: {aggregation}")
+        if src_sr != self.SAMPLE_RATE:
+            wav = torchaudio.functional.resample(wav, src_sr, self.SAMPLE_RATE)
```
Comment (GaganNarula) [inline avex/models/birdnet.py]: is this resampling step identical ? seems like your tests are ok with it

File: avex/models/birdnet.py [hunk 6]
```diff
@@ -1,574 +1,408 @@
...
+
     def idx_to_species(self, idx: int) -> str:
         return self.species[idx]
 
     def species_to_idx(self, name: str) -> int:
-        return self.species.index(name)
+        try:
+            return self.species.index(name)
+        except ValueError as err:
+            raise ValueError(f"Species '{name}' not found in BirdNet label list") from err
```
Comment (GaganNarula) [inline avex/models/birdnet.py]: LookupError would be more suitable

**Review by GaganNarula — COMMENTED** [review]
some comments

File: avex/models/birdnet.py (hunk 3 above)
Comment (nkundiushuti) [reply avex/models/birdnet.py:126]: I changed it so we have the common function for onnx models (birdnet and perch2) looking for the layer with a given name

File: avex/models/birdnet.py (hunk 4 above)
Comment (nkundiushuti) [reply avex/models/birdnet.py:77]: changed this, see the prev comment

File: avex/models/birdnet.py (hunk 5 above)
Comment (nkundiushuti) [reply avex/models/birdnet.py]: I am deferring resampling for all models (audio needs to be resampled outside)

File: avex/models/birdnet.py (hunk 2 above)
Comment (nkundiushuti) [reply avex/models/birdnet.py:113]: agree. changed it

**Review by GaganNarula — APPROVED** [review]
I thinks good to go
